pkg/config/analysis/incluster/controller.go

```diff
@@ -78,27 +80,33 @@ func NewController(stop <-chan struct{}, rwConfigStore model.ConfigStoreControll
 
 // Run is blocking
 func (c *Controller) Run(stop <-chan struct{}) {
-	t := time.NewTicker(features.AnalysisInterval)
-	oldmsgs := diag.Messages{}
-	for {
-		select {
-		case <-t.C:
-			res, err := c.analyzer.ReAnalyze(stop)
-			if err != nil {
-				log.Errorf("In-cluster analysis has failed: %s", err)
-				continue
-			}
-			// reorganize messages to map
-			index := map[status.Resource]diag.Messages{}
-			for _, m := range res.Messages {
-				key := status.ResourceFromMetadata(m.Resource.Metadata)
-				index[key] = append(index[key], m)
-			}
-			// if we previously had a message that has been removed, ensure it is removed
-			// TODO: this creates a state destruction problem when istiod crashes
-			// in that old messages may not be removed.  Not sure how to fix this
-			// other than write every object's status every loop.
-			for _, m := range oldmsgs {
+	db := concurrent.Debouncer[config.GroupVersionKind]{}
+	chKind := make(chan config.GroupVersionKind, 10)
+
+	for _, k := range c.analyzer.Schemas().All() {
+		c.analyzer.RegisterEventHandler(k.GroupVersionKind(), func(oldcfg config.Config, newcfg config.Config, ev model.Event) {
+			chKind <- oldcfg.GroupVersionKind
```

Comment:
Should we exclude some schemas or configs by default, like those leaderelection configmaps? Otherwise I think they will keep being unnecessarily analyzed since they are always changing?

Reply:
The only configmap analyzers care about today is the injectorConfig, which is only watched by one analyzer.  We could add a shortcut here that checks if the cm is an injectionConfig, and not allow others through, but that definitely leaks abstraction details.  Are there other types we should be worried about?

Reply:
I think the other one is secret. These two are the only one's I've seen recently that may result in high resource consumption.

Reply:
FYI https://github.com/istio/istio/issues/46563 and https://github.com/istio/istio/issues/45458.

Reply:
My (very unrefined) thoughts:

It feels like what we are talking about is a general issue with the analyzer architecture. We could work around it for special cases, of course -- but its not very precise.

For example, consider the "Deprecation" analyzer. Looks like this:

```go
func (fa *FieldAnalyzer) Analyze(ctx analysis.Context) {
	ctx.ForEach(gvk.VirtualService, func(r *resource.Instance) bool {
		fa.analyzeVirtualService(r, ctx)
		return true
	})
}
```
So any VS that changes, we will analyze ALL VS again (and CRD, Sidecar).

Really, though, this specific analyzer is function of `{1 VS} => []analysis`. In theory, we only need to run `fa.analyzeVirtualService(TheOneVirtualServiceThatChanged, ctx)`.

Not all analyzers are this simple though. The `ServiceAssociationAnalyzer` is more like:

```go
c.ForEach(gvk.Deployment, func(r *resource.Instance) bool {
  c.ForEach(gvk.Service, func(r *resource.Instance) bool { ... }
}
```

Even this, though, is not `AllDeployments*AllServices => []analysis`. Its really `1 Deployment * all services => []analysis`. If 1 deployment changes, we would want to re-run the analysis for that 1 deployment. If any service changes, we would need to run analysis on all deployments... with the current logic. The logic could be better though:  `1 Deployment * services selecting the deployment => []analysis`. This allows a much more precise re-run if a service changes (likely, it would be 1 deployment changing on any service).

Anyhow -- this is _exactly_ the type of problem https://docs.google.com/document/d/1h1E2EQD5aQG9cuL9qWiLYo88G5twmD66lev1Bbd0alI/edit? is perfect for solving...

An analyzer could look like:

```go
BadSelectors := NewCollection(Deployments, func(p Deployment) *Analysis {
        services := Fetch(Services, filter.Selects(d.Template.labels))
        if len(services) > 0 { return nil }
        return &Analysis{Message: "Service doesn't select anything"}
})
```

Then we can automagically get the dependency analysis.

---

Anyhow, not suggesting we block this, or even do this necessarily - just some ideas. I do think its ideal to have the analysis actually inline with the core logic, too. so it would really be something like 

```go
Endpoints :=  NewCollection(Pod, func(p Pod) (Endpoints, *Analysis) { ... }
```

And the code will both do its core logic of converting A => B, and emit any warnings along the way. Not everything is that way, of course -- some analyzers are not overlapping with the core logic, but a number of them are re-implementing the core, often incorrectly

Reply:
I've added the GVK filters used in local analysis.

Reply:
This sounds like an interesting approach to performing dependent analysis - I was thinking of the same circumstances in analysis architecture but wasn't sure of a great way to do it dependently. IMO this needs to be added to the UX milestone to make anslysis more scalable and usable in large clusters.

Reply:
@howardjohn BTW can I get access to the design doc?